bitcoin/src/pow.rs

```diff
@@ -924,9 +926,137 @@ impl kani::Arbitrary for U256 {
+///
+/// Implements the `GetNextWorkRequired` function from Bitcoin core.
+///
+/// # Args
+///
+/// `new_block_timestamp` is only used on testnet.
+///
+/// > Special difficulty rule for testnet: If the new block's timestamp is more
+/// > than 2*10 minutes then allow mining of a min-difficulty block.
+pub fn next_target<F, E>(
+    params: &Params,
+    current_height: absolute::Height,
+    current_header: block::Header,
+    new_block_timestamp: u32, // Only used on testnet for special difficulty rule.
+    mut get_block_header_by_height: F,
+) -> Result<CompactTarget, GetBlockHeaderError<E>>
+where
+    F: FnMut(absolute::Height) -> Result<block::Header, E>,
+{
+    if !is_retarget_height(current_height.inc(), params) {
+        if params.allow_min_difficulty_blocks {
+            // Special difficulty rule for testnet: If the new block's timestamp is more
+            // than 2*10 minutes then allow mining of a min-difficulty block.
+            let pow_limit = params.pow_limit.to_compact_lossy();
+            let pow_target_spacing = params.pow_target_spacing as u32;
+
+            if new_block_timestamp > current_header.time + pow_target_spacing * 2 {
+                return Ok(pow_limit);
+            } else {
+                let mut header = current_header;
+                let mut height = current_height;
+                while header.prev_blockhash != BlockHash::all_zeros()
+                    && is_retarget_height(height, params)
+                    && header.bits == pow_limit
+                {
+                    height = height.dec();
+                    header = get_block_header_by_height(height)?;
+                }
+                return Ok(header.bits);
+            }
+        }
+        return Ok(current_header.bits);
+    }
+
+    let adjustment_interval = params.difficulty_adjustment_interval() as u32;
+    // Go back by what we want to be 14 days worth of blocks
+    let height_first = current_height.saturating_sub(adjustment_interval - 1);
+    let block_first = get_block_header_by_height(height_first)?;
+
+    calculate_next_target(params, current_header, block_first.time)
+}
+
+/// Calculates the next target.
+///
+/// Implements the `CalculateNextWorkRequired` function from Bitcoin Core.
+pub fn calculate_next_target<E>(
+    params: &Params,
+    current_header: block::Header,
+    first_block_time: u32,
+) -> Result<CompactTarget, GetBlockHeaderError<E>> {
+    if params.no_pow_retargeting {
+        return Ok(current_header.bits);
+    }
+    let min_timespan = params.pow_target_timespan >> 2;
+    let max_timespan = params.pow_target_timespan << 2;
+
+    let timespan = (current_header.time - first_block_time) as u64;
+    let timespan = timespan.clamp(min_timespan, max_timespan);
+
+    let target = current_header.target();
+    let target = Target(target.0 * U256::from(timespan));
+    let target = Target(target.0 / U256::from(params.pow_target_timespan));
```

Comment:
We should keep my FIXME here, this API is ugly, and we should be able to do these operation on the `Target` itself because `U256` is private so the user API has no way to do an operation like this with the `Target`

Reply:
> this API is ugly

This is private code, its not part of the public API on purpose.

We explicitly do not want to enable arbitrary math operations on `Target`, unless I'm missing something that is the whole reason we are doing this PR?

Reply:
Exactly, we don't have the bandwidth to support `U256` as a first-class integer. (But could be interesting project if anyone has the time.)

Reply:
>We explicitly do not want to enable arbitrary math operations on Target, unless I'm missing something that is the whole reason we are doing this PR?

Mh I feel that I should be able to do simple operation on the target like `*` or `/` because we do not know how people will use the target (e.g: experimenting with something).

I understand that there is no bandwidth to support U256 math but I suspect that the same assertion will stand for supporting every use case that a person can do with a `Target`. But this is our call, someone will complain if there is some use case missing as I did

Reply:
@vincenzopalazzo we would have to implement it on top of `U256` and make it public, which we specifically don't want to. Implementing those operations on top of target doesn't make sense at all.

For experimenting, people can just copy-paste our code into their crate.